src/pfctl_worker.cpp

```diff
@@ -20,206 +20,206 @@ using namespace std;
-			return false;
-		}
-
-		// Check if master process is still alive.
-		if (getppid() != parent_pid) {
-			log(MSG_CRIT, fmt::sprintf("pfctl_worker: parent died"));
-			return false;
-		}
-
-		// If no message was received, try to receive another one.
-		if (!mq_success) {
-			continue;
-		}
-
-		assert (recvd_size == sizeof(pfctl_msg));
-
-		log(MSG_INFO, fmt::sprintf("lbpool: %s sync: start pf_table: %s", msg.pool_name, msg.table_name));
-
-		// Decode the message
-		string table_name(msg.table_name);
-		set<string> wanted_addresses;
-
-		for (int i = 0; i < MAX_NODES; i++) {
-			string wanted_address(msg.wanted_addresses[i]);
-			if (wanted_address.length())
-				wanted_addresses.insert(wanted_address);
-		}
-
-		/*
-		 * Measure total time of all pfctl operations.
-		 * Don't use std::chrono, there is a conflict with boost.
-		 */
-		chrono::high_resolution_clock::time_point t1 = chrono::high_resolution_clock::now();
-		pf_sync_table(table_name, wanted_addresses);
-		chrono::high_resolution_clock::time_point t2 = chrono::high_resolution_clock::now();
-		auto duration = chrono::duration_cast<chrono::milliseconds>(t2 - t1).count();
-		log(MSG_INFO, fmt::sprintf("lbpool: %s sync: finish pf_table: %s time: %dms", msg.pool_name, msg.table_name, duration));
-	}
-	return true;
+bool pfctl_worker_loop(message_queue *mq) {
+  pfctl_msg msg;
+  message_queue::size_type recvd_size;
+  unsigned int priority;
+  bool mq_success;
+
+  log(MSG_INFO, "pfctl_worker: entering worker loop");
+
+  running = true;
+  while (running) {
+    // Normal receive won't stop blocking when the queue
+    // is removed on master process.
+    mq_success = false;
+    try {
+      // A timed_receive will timeout after a moment so that
+      // this loop can run again and check if running was changed.
+      // That would be done by kill signal. Then this worker
+      // can gracefully terminate.
+      ptime delay = microsec_clock::universal_time() + millisec(1000);
+      mq_success = mq->timed_receive(&msg, sizeof(pfctl_msg), recvd_size,
+                                     priority, delay);
+    } catch (const exception &ex) {
+      log(MSG_CRIT,
+          fmt::sprintf("pfctl_worker: exception while receiving from queue %s",
+                       ex.what()));
+      return false;
+    }
+
+    // Check if master process is still alive.
+    if (getppid() != parent_pid) {
+      log(MSG_CRIT, fmt::sprintf("pfctl_worker: parent died"));
+      return false;
+    }
+
+    // If no message was received, try to receive another one.
+    if (!mq_success) {
+      continue;
+    }
+
+    assert(recvd_size == sizeof(pfctl_msg));
+
+    log(MSG_INFO, fmt::sprintf("lbpool: %s sync: start pf_table: %s",
+                               msg.pool_name, msg.table_name));
+
+    // Decode the message
+    string table_name(msg.table_name);
+    set<string> wanted_addresses;
+
+    for (int i = 0; i < MAX_NODES * 2; i++) {
```

Comment:
are you multiplying by 2 because of both ip protocols? i don't really get that

Reply:
Yes. It might be a leftover fromt he old code when LB nodes were duplicated. It probably should be removed.

Reply:
I checked it and this is an array of IP addresses, not LB nodes. There are up to 2 IP addresses per LB Node so the multiplication is correct.